### PR TITLE
feat(eval): loc acc evaluation

### PR DESCRIPTION
**Pull request by xhguo7**
### Description
Add localization evaluation.

### Changes
- Loc evaluation for SWE-Bench
- Output loc evaluation results through running
- Incorporate task success into post instance-level evaluation
- Optional loc evaluation

### Testing
- Runnable when setting `loc-eval` to either `True` or `False`
- Validated on `SWE-bench_Verified` test set
- Validated through output checking

### CODE REVIEW

File: openhands/core/main.py
Comment (xingyaoww) [inline openhands/core/main.py:1]: You shouldn't need to subscribe to event stream, you can simply get the last action similar to this for evaluation:

https://github.com/All-Hands-AI/OpenHands/blob/e4c284f96dc2eb7f54fd6c3c197d78f76ac14042/evaluation/benchmarks/gpqa/run_infer.py#L246-L270

File: openhands/agenthub/codeact_agent/prompts/system_prompt.j2
Comment (xingyaoww) [inline openhands/agenthub/codeact_agent/prompts/system_prompt.j2:1]: We shouldn't modify the system message here for a particular evaluation

File: evaluation/benchmarks/swe_bench/scripts/eval_infer.sh [hunk 1]
```diff
@@ -139,3 +139,38 @@ else
...
+        echo "Updating success status for $instance_id: FAILURE"
+        poetry run python evaluation/benchmarks/swe_bench/memory/update_success.py \
+            "$instance_id" --eval-dir $(dirname $LOC_EVAL_FILEPATH)
+    done
+    
+    echo "Localization success statuses updated"
+else
+    echo "Report file not found: $RESULT_OUTPUT_DIR/report.json"
+    echo "Cannot update localization success statuses"
+fi
```
Comment (xingyaoww) [inline evaluation/benchmarks/swe_bench/scripts/eval_infer.sh:176]: If localization task is so different, let's create a new folder `swe_bench_localization` with their own scripts, and only keep relevant files there (i.e., import modules/functions from `swe_bench/`)

File: evaluation/benchmarks/swe_bench/loc_eval/__init__.py
Comment (xingyaoww) [inline evaluation/benchmarks/swe_bench/loc_eval/__init__.py:1]: Can you add a README.md file under `loc_eval` folder explaining what is this folder for?

File: evaluation/benchmarks/swe_bench/loc_eval/scripts/run_infer.sh [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  EVAL_NOTE="$EVAL_NOTE-$EXP_NAME"
+fi
+# if mode != swe, add mode to the eval note
+if [ "$MODE" != "swe" ]; then
+  EVAL_NOTE="${EVAL_NOTE}-${MODE}"
+fi
+
+function run_eval() {
+  local eval_note="${1}"
+  COMMAND="poetry run python evaluation/benchmarks/swe_bench/run_infer.py \
```
Comment (xingyaoww) [inline evaluation/benchmarks/swe_bench/loc_eval/scripts/run_infer.sh:101]: Shouldn't you use the loc_eval/run_infer.py?

File: evaluation/benchmarks/swe_bench/loc_eval/scripts/run_eval_loc.sh [hunk 3]
```diff
@@ -0,0 +1,214 @@
+#!/bin/bash
+
+# Config
+INFER_DIR="./llm.claude-3-5-haiku1/litellm_proxy_claude-3-5-haiku-20241022"
+EVAL_DIR="./evaluation/evaluation_outputs/outputs/princeton-nlp__SWE-bench_Verified-test/CodeActAgent/claude-3-5-haiku-20241022_maxiter_20_N_v0.38.0-no-hint-run_1/eval_outputs"
+OUTPUT_DIR="./evaluation/benchmarks/swe_bench/loc_eval/eval_saves"
+SPLIT="test"
+DATASET="princeton-nlp/SWE-bench_Verified"
+MAX_TURN=20
```
Comment (xingyaoww) [inline evaluation/benchmarks/swe_bench/loc_eval/scripts/run_eval_loc.sh:9]: Please don't hard-code anything (providing default is ok though - similar to what we are doing in swe_bench/run_infer.sh), instead, you should document the usage of this script in README.md

File: evaluation/benchmarks/swe_bench/loc_eval/run_infer.py
Comment (xingyaoww) [inline evaluation/benchmarks/swe_bench/loc_eval/run_infer.py:1]: Please DON'T duplicate functions that already exists in `swe_bench/run_infer.py` - if you need to use it, just import relevant function from that file

File: evaluation/benchmarks/swe_bench/loc_eval/run_infer.py [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+        config.sandbox.remote_runtime_resource_factor
+    )
+
+    runtime = create_runtime(config)
+    call_async_from_sync(runtime.connect)
+
+    try:
+        initialize_runtime(runtime, instance, metadata)
+
+        message_action = get_instruction(instance, metadata)
```
Comment (xingyaoww) [inline evaluation/benchmarks/swe_bench/loc_eval/run_infer.py:166]: I thought you should modify the instruction so you simply instruct the agent to localize a particular file?

Are we still using the SWE-Bench instruction here?

File: evaluation/benchmarks/swe_bench/loc_eval/README.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            - `model_config`: the config group name for your LLM settings, as defined in your `config.toml`.
+            - `git-version`: the git commit hash of the OpenHands version you would like to evaluate.
+            - `agent`: the name of the agent for benchmarks, defaulting to `CodeActAgent`.
+            - `eval_limit`: limits the evaluation to the first `eval_limit` instances. By default, the script evaluates the entire SWE-bench_Lite test set (300 issues). 
+                - Note: in order to use `eval_limit`, you must also set `agent`.
+            - `max_iter`: the maximum number of iterations for the agent to run. By default, it is set to 100.
+            - `num_workers`: the number of parallel workers to run the evaluation. By default, it is set to 1.
+            - `dataset`: SWE-Bench dataset name. Please use huggingface format to specifies which dataset to evaluate on.
+            - `dataset_split`: SWE-Bench dataset split.
+            - `loc-eval-save`: directory to save outputs for localization
```
Comment (xingyaoww) [inline evaluation/benchmarks/swe_bench/loc_eval/README.md:26]: Why do you need an additional `loc-eval-save`? You should be able to put the result into `EvalOutput` in `run_infer.py` - so we don't need an additional dir, rather, everything would be saved into the `output.jsonl` together with instance ID

**Comment by neubig** [comment]
Hi @xhguo7 are you still working on this one?

**Comment by xhguo7** [comment]
Thank you so much for your kind reminder! I'm working on it, and will push an update soon!

**Comment by xhguo7** [comment]
Hello! I'm very sorry for the waiting!

I have updated the implementation of localization evaluation on SWE-Bench, which mainly contains the following:
- **Main code:** ./evaluation/benchmarks/swe_bench/loc_eval
- **Bash run:** ./evaluation/benchmarks/swe_bench/scripts/eval_localization.sh
- **Usage README.md:** ./evaluation/benchmarks/swe_bench/loc_eval/README.md

This implementation is completely post-processing, and has tested and validated on OpenHands (version: 0.45.0, commit: 848f692).

Thank you so much for your time and review! I would be more than happy to do any revisions if needed. Many thanks!

**Comment by mamoodi** [comment]
@xhguo7 is this ready for review? If so please click the `Ready for review` button so @neubig and @xingyaoww can be notified it needs a review.

**Comment by xhguo7** [comment]
> @xhguo7 is this ready for review? If so please click the `Ready for review` button so @neubig and @xingyaoww can be notified it needs a review.

Thank you so much for your kind reminder! My sincere apologies for the delay. I took a bit more time to test on more inference results and make some updates to better handle edge cases. It's now ready for review. Thanks again for your kind advice!

**Review by xingyaoww — APPROVED** [review]
LGTM

**Comment by xingyaoww** [comment]
@xhguo7 can you fix the linter?

**Comment by xhguo7** [comment]
> @xhguo7 can you fix the linter?

Got it! I'm working on it, and will make an update to fix this soon! Many thanks!

**Comment by mamoodi** [comment]
I don't know why it's so hard to merge this PR :D 
I'm trying to help get it merged but something is always failing. Sorry about that. Let me merge main into it one more time and see if anything improves.

**Comment by xingyaoww** [comment]
@mamoodi it finally merges! Thank you!
